src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -82,6 +82,46 @@ public partial class Activity : IDisposable
         private LinkedList<ActivityEvent>? _events;
         private Dictionary<string, object>? _customProperties;
         private string? _displayName;
+        private ActivityStatusCode _statusCode;
+        private string? _statusDescription;
+
+        /// <summary>
+        /// Gets status code of the current activity object.
+        /// </summary>
+        public ActivityStatusCode Status => _statusCode;
+
+        /// <summary>
+        /// Gets the status descrition of the current activity object.
+        /// </summary>
+        public string? StatusDescription => _statusDescription;
+
+        /// <summary>
+        /// Sets the status code and description on the current activity object.
+        /// </summary>
+        /// <param name="code">The status code</param>
+        /// <param name="description">The error status descrition</param>
+        /// <returns>'this' for convenient chaining</returns>
+        /// <remarks>
+        /// It is not allowed to set the status code to Unset value.
+        /// The description parameter is allowed to be non-null value only when passing 'Error' status code.
+        /// </remarks>
+        public Activity SetStatus(ActivityStatusCode code, string? description = null)
+        {
+            if (code <= ActivityStatusCode.Unset || code > ActivityStatusCode.Error)
+            {
+                throw new ArgumentException(SR.Format(SR.InvalidStatusCode, code), nameof(code));
```

Comment:
ArgumentOutOfRangeExcception?

Reply:
I am not sure of that if they try `Unset` value. should this case considered out of range?

Reply:
What if we let them set any value they want and don't have an error condition at all?

Reply:
Don't see any reason why we want to prevent users from calling SetStatus(ActivityStatusCode.Unset). Only restriction is about Ignoring `Description` except for `Error`.

Reply:
The Question now for @noahfalk suggestion. do you think we can allow any arbitrary values? I mean is it ok to use something like `(ActivityStatusCode) (-10)`?

@reyang as you were the one advocating for not allowing `Unset`, could you please respond if you have any objection allowing `Unset` to be set and what you think allowing any status code even not defined ones in ActivityStatusCode?

If all agrees, I can modify the code as follows: 

- Allow any status code to be set (including non defined values).
- Any code not equal `Error`, I'll reset the status description to `null`.

Reply:
I will refer to the spec https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status.
My interpretation is that:
* the default value is `unset`
* instrumentation library could set it to `error`, with an optional description
* others (e.g. application) could set it to `ok`, in this case there shouldn't be a description (so the description is set back to null)

Other cases should not happen, we can either ignore, throw exception, or fall back to a default behavior.

Reply:
I had offline discussion with @reyang and @cijothomas and here is the behavior we are going to go with:

- We'll allow setting all status codes. That include `Unset` or even undefined values (e.g. `(ActivityStatusCode)100`).
- Setting any code different than `Error` will reset the description to null.
- The description can be non null only with `Error` code.
- We'll never throw exceptions from setting the status. 

I'll go ahead and update the code according to that.